README.md
# Audio decoder for RIFF WAVE files

[![Generic badge](https://img.shields.io/badge/CurrentVersion-0.0.1-blue.svg)](.)

* extract RIFF WAVE file header that contains the format of the sound information
* read arbitrary sample block size (buffer size)

CHANGELOG.md
# Changelog for wave-decoder

2020-05-15 00:17    clean up pom.xml (Rémi Georgiou)  
2020-05-15 00:12    Initial commit (Rémi Georgiou)  


src/main/java/ch/georgiou/audio/wavedecoder/Application.java
package ch.georgiou.audio.wavedecoder;

import ch.georgiou.audio.wavedecoder.decoder.Decoder;
import ch.georgiou.audio.wavedecoder.decoder.RIFFWaveHeader;
import ch.georgiou.audio.wavedecoder.player.AudioPlayer;
import lombok.extern.log4j.Log4j2;
import org.jooq.lambda.Unchecked;

import javax.annotation.Nonnull;
import java.io.FileInputStream;
import java.util.Objects;
import java.util.function.Consumer;

@Log4j2
public class Application {

    private static final int PCM_SAMPLE_BLOCK_SIZE = 64;

    public static void main(String[] args) {
        log.debug("Available CPU cores: {}", Runtime.getRuntime().availableProcessors());

        if (args.length < 1) {
            usage();
            throw new IllegalArgumentException();
        }

        start(args[0]);
    }

    private static void start(@Nonnull String fileName) {
        AudioPlayer audioPlayer = new AudioPlayer();
        decode(Unchecked.consumer(decoder -> {
            decoder.setStream(new FileInputStream(fileName));
            RIFFWaveHeader riffWaveHeader = Objects.requireNonNull(decoder.getRIFFWaveHeader());
            log.info(riffWaveHeader);
            audioPlayer.playback(riffWaveHeader, line -> {
                byte[] sampleBlock;
                log.debug("PCM sample block size: " + PCM_SAMPLE_BLOCK_SIZE);
                log.info("starting playback...");
                while ((sampleBlock = decoder.readNextNSamples(PCM_SAMPLE_BLOCK_SIZE)) != null) {
                    line.write(sampleBlock, 0, sampleBlock.length);
                }
                log.info("playback stopped");
            });
        }));
    }

    private static void decode(@Nonnull Consumer<Decoder> decoderConsumer) {
        try (Decoder decoder = new Decoder()) {
            decoderConsumer.accept(decoder);
        }
    }


    private static void usage() {
        log.error("Program argument is missing. Please provide an absolute file path.");
    }
}


src/main/java/ch/georgiou/audio/wavedecoder/player/AudioPlayer.java
package ch.georgiou.audio.wavedecoder.player;

import ch.georgiou.audio.wavedecoder.decoder.RIFFWaveHeader;

import javax.annotation.Nonnull;
import javax.sound.sampled.AudioFormat;
import javax.sound.sampled.AudioSystem;
import javax.sound.sampled.DataLine;
import javax.sound.sampled.LineUnavailableException;
import javax.sound.sampled.SourceDataLine;
import java.util.function.Consumer;

public class AudioPlayer {

    public void playback(@Nonnull RIFFWaveHeader riffWaveHeader, @Nonnull Consumer<SourceDataLine> lineConsumer)
            throws LineUnavailableException {
        AudioFormat audioFormat = getAudioFormat(riffWaveHeader);
        try (SourceDataLine line = getSourceDataLine(audioFormat)) {
            line.open(audioFormat);
            line.start();
            lineConsumer.accept(line);
            line.drain();
            line.stop();
        }
    }

    @Nonnull
    private SourceDataLine getSourceDataLine(@Nonnull AudioFormat audioFormat) throws LineUnavailableException {
        DataLine.Info info = new DataLine.Info(SourceDataLine.class, audioFormat);
        if (!AudioSystem.isLineSupported(info)) {
            throw new RuntimeException("Audio format is not supported by audio system.");
        }
        return AudioSystem.getSourceDataLine(audioFormat);
    }

    @Nonnull
    private AudioFormat getAudioFormat(@Nonnull RIFFWaveHeader riffWaveHeader) {
        return new AudioFormat(getEncoding(riffWaveHeader),
                riffWaveHeader.getSampleRate(),
                riffWaveHeader.getBitsPerSample(),
                riffWaveHeader.getChannels(),
                riffWaveHeader.getBytesPerSample(),
                riffWaveHeader.getBytesPerSecond(),
                false);
    }

    /**
     * Inconsistency in RIFF WAVE file format.<br>
     * 8-bit samples are stored as unsigned bytes [0, 255].<br>
     * 16-bit samples are stored as 2's-complement signed shorts [-32768, 32767].
     * @param riffWaveHeader RIFF WAVE header
     * @return the audio format
     */
    @Nonnull
    private AudioFormat.Encoding getEncoding(@Nonnull RIFFWaveHeader riffWaveHeader) {
        return riffWaveHeader.getBitsPerSample() == 8 ?
                AudioFormat.Encoding.PCM_UNSIGNED : AudioFormat.Encoding.PCM_SIGNED;
    }
}


src/main/java/ch/georgiou/audio/wavedecoder/decoder/RIFFWaveHeader.java
package ch.georgiou.audio.wavedecoder.decoder;

import javax.annotation.Nonnull;
import java.util.Locale;
import java.util.Objects;

/**
 * Header of a RIFF WAVE audio file containing the format of the sound information
 *
 * @author Rémi Georgiou
 */
public class RIFFWaveHeader {

    static final int RIFF_CHUNK_DESCRIPTOR = 0x46464952;          // "RIFF" in ASCII little endian
    static final int WAVE_CHUNK_DESCRIPTOR = 0x45564157;          // "WAVE" in ASCII little endian
    static final int FORMAT_CHUNK_DESCRIPTOR = 0x20746d66;        // "fmt " in ASCII little endian
    static final int DATA_CHUNK_DESCRIPTOR = 0x61746164;          // "data" in ASCII little endian
    static final int PCM_SUB_CHUNK1_SIZE = 16;

    private final AudioCodingFormat audioCodingFormat;
    private final int channels;
    private final int sampleRate;
    private final int bitsPerSample;
    private final int dataSize;

    private volatile int hashCode;

   public RIFFWaveHeader(@Nonnull AudioCodingFormat audioCodingFormat, int channels, int sampleRate, int bitsPerSample,
                         int dataSize) {
        this.audioCodingFormat = audioCodingFormat;
        this.channels = channels;
        this.sampleRate = sampleRate;
        this.bitsPerSample = bitsPerSample;
        this.dataSize = dataSize;
    }

    /**
     * Returns the audio coding format.
     * @return  audio coding format
     */
    @Nonnull
    public AudioCodingFormat getAudioCodingFormat() {
        return audioCodingFormat;
    }

    /**
     * Returns the number of channels.
     * @return  number of channels
     */
    public int getChannels() {
        return  channels;
    }

    /**
     * Returns the sample rate.
     * @return  sample rate
     */
    public int getSampleRate() {
        return sampleRate;
    }

    /**
     * Returns the number of bits per sample.
     * @return  quantisation bits per sample
     */
    public int getBitsPerSample() {
        return bitsPerSample;
    }

    /**
     * Returns the number of bytes (for all channels) to represent one sample of data.
     * This is sometimes called the block alignment.
     * Definition: bits per sample * channels / 8
     * @return  bytes per sample
     */
    public int getBytesPerSample() {
        return bitsPerSample * channels / 8;
    }

    /**
     * Returns the number of bytes per second which is the speed of the data stream in bytes per second.
     * Definition: sample rate * bits per sample * channels / 8
     * @return  bytes per second
     */
    public int getBytesPerSecond() {
        return sampleRate * bitsPerSample * channels / 8;
    }

    /**
     * Returns the number of bytes of PCM data that is included in the data section.
     * @return  the actual PCM audio data size in bytes
     */
    public int getDataSize() {
        return dataSize;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }

        if (!(o instanceof RIFFWaveHeader)) {
            return false;
        }

        RIFFWaveHeader riffWaveHeader = (RIFFWaveHeader) o;
        return channels == riffWaveHeader.channels
                && sampleRate == riffWaveHeader.sampleRate
                && bitsPerSample == riffWaveHeader.bitsPerSample
                && dataSize == riffWaveHeader.dataSize
                && audioCodingFormat == riffWaveHeader.audioCodingFormat;
    }

    @Override
    public int hashCode() {
        if (hashCode == 0) {
            hashCode = Objects.hash(audioCodingFormat, channels, sampleRate, bitsPerSample, dataSize);
        }
        return hashCode;
    }

    @Override
    public String toString() {
        return String.format(Locale.getDefault(),
                "RIFF WAVE header:  audio coding format=%s  channels=%d  sample rate=%d Hz  bits per sample=%d"
                + "  block alignment=%d bytes  data size=%d bytes",
                AudioCodingFormat.LINEAR_PCM.name(), channels, sampleRate, bitsPerSample,
                bitsPerSample * channels / 8, dataSize);
    }
}


src/main/java/ch/georgiou/audio/wavedecoder/decoder/AudioCodingFormat.java
package ch.georgiou.audio.wavedecoder.decoder;

import javax.annotation.Nonnull;
import java.util.Arrays;

public enum AudioCodingFormat {

    LINEAR_PCM(1, "linear pulse-code modulation");

    private final int code;
    private final String name;

    AudioCodingFormat(int code, @Nonnull String name) {
        this.code = code;
        this.name = name;
    }

    public int getCode() {
        return code;
    }

    @Nonnull
    public String getName() {
        return name;
    }

    @Nonnull
    public static AudioCodingFormat from(int value) {
        return Arrays.stream(values())
                .filter(audioCodingFormat -> audioCodingFormat.code == value)
                .findFirst()
                .orElseThrow(() -> new RuntimeException("Unknown audio coding format code provided: " + value));
    }
}


src/main/java/ch/georgiou/audio/wavedecoder/decoder/Decoder.java
package ch.georgiou.audio.wavedecoder.decoder;

import lombok.extern.log4j.Log4j2;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.io.IOException;
import java.io.InputStream;

/**
 * <p>This decoder handles uncompressed linear PCM audio files, the most common type of RIFF (Resource Interchange File Format) files.
 * The header contains the format of the sound file and provides therefore information about file type, audio coding format,
 * sample rate, quantisation bits per sample, number of channels, byte rate, as well as the length of audio data.</p>
 * <p>The header of a RIFF WAVE file is 44 bytes in size and has the following format:</p>
 *
 * <table>
 *     <caption> </caption>
 * <thead>
 * <tr>
 *      <th>Position</th>
 *      <th>Example value</th>
 *      <th>Description</th>
 * </tr>
 * </thead>
 * <tr>
 *      <td>0 - 3</td>
 *      <td>"RIFF"</td>
 *      <td>RIFF chunk descriptor containing the characters "RIFF" in ASCII. Marks the file as a RIFF file container. 0x46464952 in little endian.</td>
 * </tr>
 * <tr>
 *      <td>4 - 7</td>
 *      <td>6991680</td>
 *      <td>RIFF chunk size representing the overall file size - 8 bytes (RIFF chunk descriptor and this value).</td>
 * </tr>
 * <tr>
 *      <td>8 - 11</td>
 *      <td>"WAVE"</td>
 *      <td>WAVE format descriptor containing the characters "WAVE" in ASCII. 0x45564157 in little endian.</td>
 * </tr>
 * <tr>
 *      <td>12 - 15</td>
 *      <td>"fmt "</td>
 *      <td>Format sub-chunk descriptor containing the characters "fmt " (including trailing whitespace) describing the sound format. 0x20746d66 in little endian.</td>
 * </tr>
 * <tr>
 *      <td>16 - 19</td>
 *      <td>16</td>
 *      <td>Length of format data as listed above. This value is always 16 for a RIFF WAVE audio file.</td>
 * </tr>
 * <tr>
 *      <td>20 - 21</td>
 *      <td>1</td>
 *      <td>Audio coding format. E.g. 1 is linear pulse-code modulation.</td>
 * </tr>
 * <tr>
 *      <td>22 - 23</td>
 *      <td>2</td>
 *      <td>Number of channels.</td>
 * </tr>
 * <tr>
 *      <td>24 - 27</td>
 *      <td>48000</td>
 *      <td>Sample rate in Hz (number of samples per second)</td>
 * </tr>
 * <tr>
 *      <td>28 - 31</td>
 *      <td>176400</td>
 *      <td>The byte rate is the speed of the audio data stream. Sample Rate * bits per sample * channels / 8</td>
 * </tr>
 * <tr>
 *      <td>32 - 33</td>
 *      <td>4</td>
 *      <td>Block alignment expressed in bytes. The number of bytes for one sample including all channels.</td>
 * </tr>
 * <tr>
 *      <td>34 - 35</td>
 *      <td>16</td>
 *      <td>Quantisation bits per sample</td>
 * </tr>
 * <tr>
 *      <td>36 - 39</td>
 *      <td>"data"</td>
 *      <td>data sub-chunk descriptor containing the characters "data". 0x61746164 in little endian. Marks the beginning of the data section.</td>
 * </tr>
 * <tr>
 *      <td>40 - 43</td>
 *      <td>6991448</td>
 *      <td>Size of the audio data in bytes.</td>
 * </tr>
 * </table>
 *
 * <p>8-bit audio is stored as unsigned bytes in the range [0, 255].<br>
 *    16-bit audio is usually represented as signed 2s'-complement short values in the range [-32768, 32767]</p>
 *
 * @author Rémi Georgiou
 */
@Log4j2
public class Decoder implements AutoCloseable {

    private InputStream audioStream;
    private RIFFWaveHeader riffWaveHeader;

    /**
     * Initialise the decoder with the {@link InputStream} of a RIFF WAVE sound file.<br>
     * @param audioStream the audio stream
     */
    public synchronized void setStream(@Nonnull InputStream audioStream) {
        this.audioStream = audioStream;
    }

    /**
     * Reads a sample block of size N samples or the minimum of available bytes of the audio stream.<br>
     * This method returns N samples <b>including all channels.</b><br>
     * Returns null when the end of the stream is reached.
     * @param samples the number of samples to be read from the audio stream.
     * @return A sample block of size N samples.
     */
    @Nullable
    public byte[] readNextNSamples(int samples) {
        try {
            if (audioStream == null || audioStream.available() == 0) {
                log.info("reached end of stream");
                return null;
            }
            byte[] sampleBlock = new byte[Math.min(riffWaveHeader.getChannels() * samples, audioStream.available())];
            audioStream.read(sampleBlock);
            return sampleBlock;
        } catch (IOException e) {
            log.error(e);
        }
        return null;
    }

    /**
     * The decoder attempts to extract the RIFF WAVE file header from the audio stream.
     * @return The RIFF WAVE file header
     */
    @Nonnull
    public RIFFWaveHeader getRIFFWaveHeader() {
        if (riffWaveHeader == null) {
            riffWaveHeader = extractRIFFWaveHeader();
        }
        return riffWaveHeader;
    }

    /**
     * Close the underlying audio stream of the decoder instance.
     */
    @Override
    public void close() {
        try {
            audioStream.close();
        } catch (IOException e) {
            log.error("failed to close audio stream");
        }
    }

    /**
     * Read first 4 bytes from the byte array and transform it to a 32-bit integer in little endian form.
     * @param bytes a byte array
     * @return integer value
     */
    int getIntLittleEndian(@Nonnull byte[] bytes) {
        return (bytes[0] & 0xff)
                | ((bytes[1] & 0xff) << 8)
                | ((bytes[2] & 0xff) << 16)
                | ((bytes[3] & 0xff) << 24);
    }

    /**
     * Read first 2 bytes from the byte array and transform it to a 16-bit short in little endian form.
     * @param bytes a byte array
     * @return short value
     */
    short getShortLittleEndian(@Nonnull byte[] bytes) {
        return (short) ((bytes[0] & 0xff) | ((bytes[1] & 0xff) << 8));
    }

    /**
     * Extract the RIFF WAVE file header from the audio stream containing the format of the sound information
     * @return the RIFF WAVE file header
     */
    @Nonnull
    private RIFFWaveHeader extractRIFFWaveHeader() {
        try {
            // Check if file container format is RIFF (Resource Interchange File Format).
            int riffChunkId = getIntLittleEndian(audioStream.readNBytes(4));
            if (riffChunkId != RIFFWaveHeader.RIFF_CHUNK_DESCRIPTOR) {
                throw new DecoderException("Not a RIFF file.");
            }

            // This is the file size minus 8 bytes (RIFF chunk ID and RIFF chunk size) - ignored
            int chunkSize = getIntLittleEndian(audioStream.readNBytes(4));

            // Check if the file format is WAVE.
            int format = getIntLittleEndian(audioStream.readNBytes(4));
            if (format != RIFFWaveHeader.WAVE_CHUNK_DESCRIPTOR) {
                throw new DecoderException("Unknown file format - not WAVE.");
            }

            // Check the format sub-chunk.
            int formatChunkId = getIntLittleEndian(audioStream.readNBytes(4));
            if (formatChunkId != RIFFWaveHeader.FORMAT_CHUNK_DESCRIPTOR) {
                throw new DecoderException("Illegal format sub-chunk.");
            }

            // Size of the PCM sub-chunk - should be 16 bytes.
            int pcmSubChunkSize = getIntLittleEndian(audioStream.readNBytes(4));
            if (pcmSubChunkSize != RIFFWaveHeader.PCM_SUB_CHUNK1_SIZE) {
                throw new DecoderException("Illegal PCM sub-chunk size: " + pcmSubChunkSize);
            }

            // Read two bytes at position 20 and check if the audio coding format is linear PCM.
            short encodingFormat = getShortLittleEndian(audioStream.readNBytes(2));
            if (encodingFormat != AudioCodingFormat.LINEAR_PCM.getCode()) {
                throw new DecoderException("Unsupported audio coding format");
            }
            log.debug("audio coding format: " + encodingFormat);

            // Read two bytes at position 22 to get the number of channels.
            short channels = getShortLittleEndian(audioStream.readNBytes(2));
            log.debug("number of channels: " + channels);

            // Read four bytes at position 24 and to get the sample rate.
            int sampleRate = getIntLittleEndian(audioStream.readNBytes(4));
            log.debug("sample rate: " + sampleRate);

            /*
                Read four bytes at position 28 to get the byte rate.
                Bytes per second is the speed of the audio data stream:
                sample rate * number of channels * bits per sample / 8
             */
            int bytesPerSecond = getIntLittleEndian(audioStream.readNBytes(4));
            log.debug("audio data stream speed: {} bytes/second", bytesPerSecond);

            /*
                Read two bytes at position 32 to get the block alignment
                (the number of bytes for one sample including all channels.)
             */
            short blockAlignment = getShortLittleEndian(audioStream.readNBytes(2));
            log.debug("block alignment: " + blockAlignment);

            // Read two bytes at position 34 to get the quantisation bits per sample
            short bitsPerSample = getShortLittleEndian(audioStream.readNBytes(2));
            log.debug("quantisation bits per sample: " + bitsPerSample);

            /*
                In general the RIFF WAVE header size is 44 bytes. There may be additional sub-chunks. If so,
                each will have a 4 bytes SubChunkID, 4 bytes of SubChunkSize and SubChunkSize amount of data.
                The rest is audio data.
                "data" marker should follow at position 36 - skip over any padding/junk data.
             */
            int junkData = 0;
            while (getIntLittleEndian(audioStream.readNBytes(4)) != RIFFWaveHeader.DATA_CHUNK_DESCRIPTOR) {
                int subChunkSize = getIntLittleEndian(audioStream.readNBytes(4));
                junkData += audioStream.skip(subChunkSize);
            }
            log.debug("skipped over {} bytes of junk data", junkData);

            int dataSize = getIntLittleEndian(audioStream.readNBytes(4));
            return new RIFFWaveHeader(AudioCodingFormat.from(encodingFormat), channels, sampleRate, bitsPerSample, dataSize);
        } catch (IOException | DecoderException ex) {
            log.error("Error while extracting RIFF WAVE header. " + ex.getMessage());
            throw new RuntimeException();
        }
    }
}
